esileapclient/tests/functional/base.py

```diff
@@ -0,0 +1,173 @@
+            # 3) from the environment variables (e.g. OS_PASSWORD, etc)
+            for opt in ('username', 'password', 'project_name', 'auth_type',
+                        'auth_url', 'user_domain_name', 'project_domain_name'):
+                for sect in 'keystone', 'keystone_authtoken':
+                    if cfg_parser.has_option(sect, opt):
+                        auth_opts[opt] = cfg_parser.get(sect, opt)
+                        break
+                if opt not in auth_opts.keys():
+                    x = os.environ.get('OS_%s' % opt.upper())
+                    if x is not None:
+                        auth_opts[opt] = x
+                    else:
+                        raise configparser.NoOptionerror
+        except (configparser.NoOptionError):
+            self.fail("Missing option %s in configuration file." % opt)
+
+        self.config['auth_type'] = auth_opts['auth_type']
+        self.config['auth_url'] = auth_opts['auth_url']
+
+        self.users['admin'] = {
+            'name': auth_opts['username'],
+            'password': auth_opts['password'],
+            'domain': auth_opts['user_domain_name']
+        }
+        self.projects['admin'] = {
+            'name': auth_opts['project_name'],
+            'domain': auth_opts['project_domain_name']
+        }
+
+    def _init_client(self, user, project):
+        self.clients[user] = (
+            ESICLIClient(cli_dir=self.config['cli_dir'],
+                         username=self.users[user]['name'],
+                         password=self.users[user]['password'],
+                         user_domain_name=self.users[user]['domain'],
+                         tenant_name=self.projects[project]['name'],
+                         project_domain_name=self.projects[project]['domain'],
+                         identity_api_version='3',
+                         uri=self.config['auth_url']))
+
+    def _init_dummy_project(self, name, roles, parent=None):
+        admin_client = self.clients['admin']
+
+        project_name = data_utils.rand_name('esi-%s-project' % name)
+        flags = '--domain default --enable -f json'
+        if parent:
+            if parent not in self.projects.keys():
+                raise NameError('Invalid parent project: %s' % parent)
+            flags += ' --parent %s' % self.projects[parent]['name']
+        output = admin_client.openstack('project create %s' % project_name,
+                                        '', flags)
+        self._cls_cleanups.append([admin_client.openstack, 'project delete',
+                                   '', '%s' % project_name])
+        project_id = json.loads(output)['id']
+        self.projects[name] = {
+            'name': project_name,
+            'id': project_id,
+            'parent': parent,
+            'domain': 'default'
+        }
+
+        if roles is [] or type(roles) is not list:
+            raise ValueError('Invalid role list for dummy project %s' % name)
```

Comment:
I think this type of check is technically not necessary; if you pass in something that's not a list, the code will simply throw a complaint about 'roles being non iterable' in the next line.

If you feel like it's absolutely necessary to check the type, I think ``isinstance`` is the more accepted way to go:

https://stackoverflow.com/questions/43106905/what-is-the-best-way-to-check-if-a-variable-is-a-list